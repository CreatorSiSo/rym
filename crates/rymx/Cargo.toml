[package]
name = "rymx"
version = "0.0.3"
edition = "2021"
default-run = "rymx"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
anyhow = "1.0"
clap = { version = "4.4", features = ["cargo"] }
insta = { version = "1.31.0", features = ["glob"] }
logos = "0.13.0"
owo-colors = "3.5"
rustyline = "12"

[features]
expand = []

[profile.dev.package.insta]
opt-level = 3

[profile.dev.package.similar]
opt-level = 3
