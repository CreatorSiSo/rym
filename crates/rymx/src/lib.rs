use std::fmt::Write;

mod ast;
mod compile;
mod diagnostics;
mod interpret;
mod parse;
mod span;
pub mod std_lib;
mod tokenize;

pub use diagnostics::{Diagnostics, SourceId};
pub use interpret::Env;
pub use span::Span;
pub use tokenize::tokenizer;

use ast::{Expr, Module};
use interpret::{Interpret, Value};
use span::SourceSpan;
use tokenize::Token;

pub fn interpret(diag: &mut Diagnostics, env: &mut Env, ast: impl Interpret) -> Value {
	diag.start_stage("interpret");
	// TODO does this make sense
	// Ignoring control flow
	let result = ast.eval(env).inner();

	let env_state: String = env
		.variables()
		.into_iter()
		.fold((0, String::new()), |(indent, mut accum), scope| {
			for (name, (kind, value)) in scope {
				writeln!(accum, "{}{kind} {name} = {}", "  ".repeat(indent), value)
					.expect("Internal Error: Unable to write into String");
			}
			(indent + 1, accum)
		})
		.1;
	diag.push_result(&env_state);

	result
}

pub fn compile_module(diag: &mut Diagnostics, src: &str, src_id: SourceId) -> Option<Module> {
	diag.start_stage("tokenize");
	let tokens: Vec<(Token, Span)> = tokenize(diag, src, src_id.clone())?;

	diag.start_stage("parse");
	let module = match parse::parse_file(&tokens, src, src_id) {
		Ok(module) => module,
		Err(reports) => {
			for report in reports {
				diag.push_report(report);
			}
			return None;
		}
	};
	diag.push_result(&format!("{module:#?}\n"));

	// TODO Name resolution
	// TODO Typechecking
	// TODO Const evaluation
	// TODO Generate intermediate representation

	// i / / / ;

	Some(module)
}

// TODO take a module (for name lookup and so on) as input
pub fn compile_expr(diag: &mut Diagnostics, src: &str, src_id: SourceId) -> Option<Expr> {
	diag.start_stage("tokenize");
	let tokens: Vec<(Token, Span)> = tokenize(diag, src, src_id.clone())?;

	diag.start_stage("parse");
	let expr = match parse::parse_expr(&tokens, src, src_id) {
		Ok(expr) => expr,
		Err(reports) => {
			for report in reports {
				diag.push_report(report);
			}
			return None;
		}
	};
	diag.push_result(&format!("{expr:#?}\n"));

	Some(expr)
}

fn tokenize(diag: &mut Diagnostics, src: &str, src_id: SourceId) -> Option<Vec<(Token, Span)>> {
	let results: Vec<_> = tokenizer(src).collect();
	diag.push_result(
		&results.iter().fold(String::new(), |accum, (result, span)| {
			let (token, span) = match result {
				Ok(token) => (format!("{token:?}"), span),
				Err(_) => ("Error".into(), span),
			};
			accum + &format!("{token} [{}]\n", span.src(src).escape_debug())
		}),
	);

	let mut tokens = vec![];
	for (result, span) in results {
		let Ok(token) = result else {
			diag.error(
				format!("Invalid character [{}]", span.src(src)),
				SourceSpan(src_id, span),
			);
			return None;
		};
		match token {
			Token::DocComment | Token::Comment | Token::VSpace | Token::HSpace => continue,
			_ => tokens.push((token, span)),
		}
	}
	Some(tokens)
}
