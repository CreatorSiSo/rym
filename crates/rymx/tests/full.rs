use insta::assert_snapshot;
use rymx::Diagnostics;

#[test]
fn main() {
	insta::glob!("**/*.rym", |path| {
		let src = std::fs::read_to_string(path).unwrap();
		let mut diag = Diagnostics::new(Some(path.into()), Box::new(std::io::stderr()));
		let _ = rymx::compile_module(&mut diag, &src);
		assert_snapshot!(diag.dump_stages());
	})
}
