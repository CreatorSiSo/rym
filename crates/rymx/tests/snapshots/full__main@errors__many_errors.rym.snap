---
source: crates/rymx/tests/full.rs
expression: "String::from_utf8(out).unwrap()"
input_file: crates/rymx/tests/errors/many_errors.rym
---
Debug: Finished tokenizing
Const [const]
HSpace [ ]
Ident [test]
Colon [:]
HSpace [ ]
Union [union]
HSpace [ ]
BracketOpen [[]
String [\"test\"]
BracketClose []]
Ident [u8]
HSpace [ ]
Pipe [|]
HSpace [ ]
BracketOpen [[]
Float [1.0]
BracketClose []]
Ident [usihf]
HSpace [ ]
Assign [=]
HSpace [ ]
Int [0]
Semi [;]
VSpace [\n]

Error: Expected integer, found string
   ╭─[/home/simon/dev/rym/crates/rymx/tests/errors/many_errors.rym:1:20]
   │
 1 │ const test: union ["test"]u8 | [1.0]usihf = 0;
   │                    ───┬──  
   │                       ╰──── Expected integer
───╯

Error: Expected integer, found float
   ╭─[/home/simon/dev/rym/crates/rymx/tests/errors/many_errors.rym:1:33]
   │
 1 │ const test: union ["test"]u8 | [1.0]usihf = 0;
   │                                 ─┬─  
   │                                  ╰─── Expected integer
───╯

Debug: Finished parsing
Some(
    Module {
        name: "",
        constants: [
            (
                "test",
                Union(
                    [
                        Array(
                            Unknown,
                            Path(
                                Path(["u8"]),
                            ),
                        ),
                        Array(
                            Unknown,
                            Path(
                                Path(["usihf"]),
                            ),
                        ),
                    ],
                ),
                Literal(Int: 0),
            ),
        ],
        types: [],
        sub_modules: [],
    },
)


